File: py/searching/binary_search.py

def binary_search(arr: list, lower: int, upper: int, target: int):
    if upper-lower >= 0:
        mid = (upper+lower)//2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            return binary_search(arr, mid+1, upper, target)
        elif arr[mid] > target:
            return binary_search(arr, lower, mid-1, target)
    return -1





def main():
    size: int = int(input())
    arr: list = list(map(int, input().split(' ')))
    assert size == len(arr)
    print(binary_search(arr, 0, size-1, arr[4]), end= ' ')


if __name__ == '__main__':
    main()

File: c/source/test/binary_search.c
#include <stdio.h>
#include <stdlib.h>


int binary_search(int *arr, int size, float target){
    int lower = 0, upper = size-1, mid;
    while (upper >= lower){
        mid = (upper + lower)/2;
        if (*(arr + mid) == target){
            return mid;
        }else if (*(arr + mid) > target){
            upper = mid - 1;
        }else if (*(arr + mid) < target){
            lower = mid + 1;
        }
    }
    return -1;
}

int *generate_arr(int size){
    int *arr = (int*)malloc(size * sizeof(int));
    for (int i = 0; i < size; i++){
        *(arr + i) = i+69;
    }
    return arr;
}


int main(int argc, char **argv){
    if (argc != 3){
        printf("Usage: %s <size> <target>\n", argv[0]);
        return 1;
    }
    int size = atoi(argv[1]), target = atof(argv[2]), *arr = generate_arr(size);
    printf("%d\n", binary_search(arr, size, target));
    free(arr);
    return 0;
}
